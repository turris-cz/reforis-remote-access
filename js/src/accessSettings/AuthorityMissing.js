/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import { Button, useAlert, useAPIPost, API_STATE } from "foris";

import API_URLs from "API";

AuthorityMissing.propTypes = {
    onAuthoritySuccess: PropTypes.func.isRequired,
};

export default function AuthorityMissing({ onAuthoritySuccess }) {
    const [setAlert] = useAlert();

    const [generateResponse, generateCA] = useAPIPost(API_URLs.authority);
    useEffect(() => {
        if (generateResponse.state === API_STATE.SUCCESS) {
            onAuthoritySuccess();
        } else if (generateResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot generate certificate authority."));
        }
    }, [generateResponse, onAuthoritySuccess, setAlert]);

    return (
        <>
            <h3>{_("No certificate authority")}</h3>
            <p>
                {_(
                    "Currently there is no certificate authority (CA) dedicated to remote access. A CA is required to generate access tokens to authenticate. To proceed you need to generate it first."
                )}
            </p>
            <Button
                onClick={() => generateCA()}
                loading={generateResponse.state === API_STATE.SENDING}
                forisFormSize
            >
                {_("Generate certificate authority")}
            </Button>
        </>
    );
}
