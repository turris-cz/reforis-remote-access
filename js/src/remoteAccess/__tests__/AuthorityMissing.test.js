/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, getByText, fireEvent, wait } from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from 'jest-mock-axios';

import AuthorityMissing from "../AuthorityMissing";

describe("<AuthorityMissing />", () => {
    let componentContainer;
    const onSuccess = jest.fn();

    function generateCA() {
        fireEvent.click(getByText(componentContainer, "Generate certificate authority"));
    }

    beforeEach(() => {
        const { container } = render(<AuthorityMissing onAuthoritySuccess={onSuccess} />);
        componentContainer = container;
    });

    it("should match snapshot", () => {
        expect(componentContainer).toMatchSnapshot();
    });

    it("should send request when button is clicked", () => {
        generateCA();
        expect(mockAxios.post).toBeCalledWith("/reforis/subordinates/api/authority", undefined, expect.anything());
    });

    it("should display spinner when sending request", () => {
        generateCA();
        expect(componentContainer).toMatchSnapshot();
    });

    it("should handle error", async () => {
        generateCA();
        mockJSONError();
        await wait(() => {
            expect(mockSetAlert).toHaveBeenCalledWith("Cannot generate certificate authority");
        });
    });

    it("should handle success", async () => {
        generateCA();
        mockAxios.mockResponse({});
        await wait(() => {
            expect(onSuccess).toHaveBeenCalledTimes(1);
        });
    });
});
