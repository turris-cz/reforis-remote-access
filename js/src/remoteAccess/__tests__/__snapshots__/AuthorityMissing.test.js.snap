// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthorityMissing /> should match snapshot 1`] = `
<div>
  <h3>
    No certification authority
  </h3>
  <p>
    Currently there is no certificate authority (CA) dedicated to remote access. A CA is required to generate access tokens to authenticate. To proceed you need to generate it first.
  </p>
  <button
    class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
    type="button"
  >
     
     
    Generate CA
  </button>
</div>
`;
